src/Evolution/Particles/MonteCarlo/Actions/FluidCouplingAction.hpp

```diff
@@ -0,0 +1,122 @@
+#include "Parallel/AlgorithmExecution.hpp"
+#include "Parallel/GlobalCache.hpp"
+
+namespace Particles::MonteCarlo {
+
+/// Mutator adding the Monte-Carlo contribution
+/// to the evolution of the fluid.
+struct FluidCouplingMutator {
+  static const size_t Dim = 3;
+
+  // We modify the fluid evolved variables, and reset the coupling
+  // terms to zero.
+  using return_tags =
+      tmpl::list<grmhd::ValenciaDivClean::Tags::TildeTau,
+                 grmhd::ValenciaDivClean::Tags::TildeYe,
+                 grmhd::ValenciaDivClean::Tags::TildeS<Frame::Inertial>,
+                 Particles::MonteCarlo::Tags::CouplingTildeTau<DataVector>,
+                 Particles::MonteCarlo::Tags::CouplingTildeRhoYe<DataVector>,
+                 Particles::MonteCarlo::Tags::CouplingTildeS<DataVector, Dim>>;
+  using argument_tags = tmpl::list<
+      evolution::dg::subcell::Tags::Mesh<Dim>,
+      evolution::dg::subcell::fd::Tags::DetInverseJacobianLogicalToInertial>;
+
+  static void apply(
+      const gsl::not_null<Scalar<DataVector>*> tilde_tau,
+      const gsl::not_null<Scalar<DataVector>*> tilde_ye,
+      const gsl::not_null<tnsr::i<DataVector, Dim>*> tilde_s,
+      const gsl::not_null<Scalar<DataVector>*> coupling_tilde_tau,
+      const gsl::not_null<Scalar<DataVector>*> coupling_tilde_rho_ye,
+      const gsl::not_null<tnsr::i<DataVector, Dim>*> coupling_tilde_s,
+      const Mesh<Dim>& mesh,
+      const Scalar<DataVector>& det_inverse_jacobian_logical_to_inertial) {
+    // Coupling terms are on the mesh with ghost zones, while the evolved
+    // variables are not. We also need to normalize the coupling terms with the
+    // cell volume to get the evolution of energy/momentum density
+    const Index<3>& extents = mesh.extents();
+    const size_t num_ghost_zones = 1;
+
+    Scalar<DataVector> det_jacobian_logical_to_inertial(*tilde_tau);
+    get(det_jacobian_logical_to_inertial) =
+        1.0 / get(det_inverse_jacobian_logical_to_inertial);
+    Scalar<DataVector> cell_inertial_three_volume =
+        make_with_value<Scalar<DataVector>>(*tilde_tau, 0.0);
+    cell_inertial_coordinate_three_volume_finite_difference(
+        &cell_inertial_three_volume, mesh, det_jacobian_logical_to_inertial);
+
+    for (size_t i = 0; i < extents[0]; i++) {
+      for (size_t j = 0; j < extents[1]; j++) {
+        for (size_t k = 0; k < extents[2]; k++) {
+          // The coupling terms are computed on a grid with ghost zone points,
+          // the fluid variables are on the grid without GZ points
+          // Index without GZ
+          const size_t local_idx =
+              i + j * extents[0] + k * extents[0] * extents[1];
+          // Index with GZ
+          const size_t extended_idx =
+              num_ghost_zones + i +
+              (num_ghost_zones + j) * (extents[0] + 2 * num_ghost_zones) +
+              (num_ghost_zones + k) * (extents[0] + 2 * num_ghost_zones) *
+                  (extents[1] + 2 * num_ghost_zones);
```

Comment:
Do or should we factor out these computations into a free function in a header file? That way if they get used in multiple places we only need to update one location and it gets updated everywhere.